src/Microsoft.AspNet.Routing/InlineRouteParameterParser.cs

```diff
@@ -12,82 +9,230 @@ namespace Microsoft.AspNet.Routing
+                }
+                else if ((currentChar == ':' || currentChar == '=') && startIndex != currentIndex)
+                {
+                    // Parameter names are allowed to start with delimiters used to denote constraints or default values.
+                    // i.e. "=foo" or ":bar" would be treated as parameter names rather than default value or constraint
+                    // specifications.
+                    parameterName = routeParameter.Substring(startIndex, currentIndex - startIndex);
+
+                    // Roll the index back and move to the constraint parsing stage.
+                    currentIndex--;
+                    break;
+                }
+
+                currentIndex++;
             }
 
-            var parameterName = parameterMatch.Groups["parameterName"].Value;
+            var parseResults = ParseConstraints(routeParameter, currentIndex, endIndex);
+            currentIndex = parseResults.CurrentIndex;
 
-            // Add the default value if present
-            var defaultValueGroup = parameterMatch.Groups["defaultValue"];
-            var defaultValue = GetDefaultValue(defaultValueGroup);
-
-            // Register inline constraints if present
-            var constraintGroup = parameterMatch.Groups["constraint"];
-            var inlineConstraints = GetInlineConstraints(constraintGroup);
+            string defaultValue = null;
+            if (currentIndex <= endIndex &&
+                routeParameter[currentIndex] == '=')
+            {
+                defaultValue = routeParameter.Substring(currentIndex + 1, endIndex - currentIndex);
+            }
 
             return TemplatePart.CreateParameter(parameterName,
                                                 isCatchAll,
                                                 isOptional,
                                                 defaultValue,
-                                                inlineConstraints);
+                                                parseResults.Constraints);
         }
 
-        private static string GetDefaultValue(Group defaultValueGroup)
+        private static ConstraintParseResults ParseConstraints(
+            string routeParameter,
+            int currentIndex,
+            int endIndex)
         {
-            if (defaultValueGroup.Success)
+            var inlineConstraints = new List<InlineConstraint>();
+            var state = ParseState.Start;
+            var startIndex = currentIndex;
+            do
             {
-                var defaultValueMatch = defaultValueGroup.Value;
-
-                // Strip out the equal sign at the beginning
-                Debug.Assert(defaultValueMatch.StartsWith("=", StringComparison.Ordinal));
-                return defaultValueMatch.Substring(1);
-            }
-
-            return null;
+                var currentChar = currentIndex > endIndex ? char.MinValue : routeParameter[currentIndex];
```

Comment:
An alternative might be to use `-1` in line with [`StringReader.Read`](https://msdn.microsoft.com/en-us/library/ath1fht8\(v=vs.110\).aspx)